### PR TITLE
Feat: Per-Packet Contexts (#62)

### LINKED ISSUE #62: Frisbee Comment and Protocol update
Frisbee needs updates on its comments around the new packet changes, and the protocol package must be refactored to reflect our packet.Metadata semantics. 

Moreover, we should fix low-hanging bugs in the heartbeat semantics and how the client and server pass messages around. 

Finally, we should also add the ability to set a per-packet context - for things like setting deadlines for certain functions.

### PR DESCRIPTION
**Pull request by ShivanshVij**
## Description

Please include a summary of the change and which issue has been fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #62

## Type of change

**Please update the title of this PR to reflect the type of change.** (Delete the ones that aren't required).

- [x] Bug fix (non-breaking change which fixes an issue) [title: 'fix:']
- [x] New feature (non-breaking change which adds functionality) [title: 'feat:']
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Testing

Please make sure that existing test cases pass (with `go test ./... -race` and `go test -bench=. ./... -race`), 
and if required please add new test cases and list them below:

## Benchmarking

Frisbee tries to adhere to strict performance requirements, so please make sure to run `go test -bench=. ./...` and paste the results below:

### Benchmarking Results:

```shell
goos: darwin
goarch: arm64
pkg: github.com/loopholelabs/frisbee
BenchmarkAsyncThroughputPipe/32_Bytes-10                  129174              9335 ns/op         342.79 MB/s         448 B/op          8 allocs/op
BenchmarkAsyncThroughputPipe/512_Bytes-10                  74641             15778 ns/op        3244.95 MB/s         449 B/op          8 allocs/op
BenchmarkAsyncThroughputPipe/1024_Bytes-10                 61093             19445 ns/op        5266.20 MB/s         451 B/op          8 allocs/op
BenchmarkAsyncThroughputPipe/2048_Bytes-10                 44413             27188 ns/op        7532.81 MB/s         458 B/op          8 allocs/op
BenchmarkAsyncThroughputPipe/4096_Bytes-10                 27327             43794 ns/op        9352.96 MB/s         476 B/op          8 allocs/op
BenchmarkAsyncThroughputNetwork/32_Bytes-10                45952             25034 ns/op         127.83 MB/s         256 B/op          4 allocs/op
BenchmarkAsyncThroughputNetwork/512_Bytes-10               39771             29964 ns/op        1708.75 MB/s         257 B/op          4 allocs/op
BenchmarkAsyncThroughputNetwork/1024_Bytes-10              32985             35735 ns/op        2865.55 MB/s         257 B/op          4 allocs/op
BenchmarkAsyncThroughputNetwork/2048_Bytes-10              24039             49823 ns/op        4110.56 MB/s         257 B/op          4 allocs/op
BenchmarkAsyncThroughputNetwork/4096_Bytes-10              17014             71983 ns/op        5690.21 MB/s         260 B/op          4 allocs/op
BenchmarkThroughputClient/test-10                             84          13213917 ns/op        2539.29 MB/s       22609 B/op         54 allocs/op
BenchmarkThroughputResponseClient/test-10                     69          14703612 ns/op        2282.02 MB/s       17730 B/op         49 allocs/op
BenchmarkThroughputServer/test-10                            100          13113973 ns/op        2558.64 MB/s        4723 B/op         12 allocs/op
BenchmarkThroughputResponseServer/test-10                    100          13495682 ns/op        2486.27 MB/s       21683 B/op         60 allocs/op
BenchmarkSyncThroughputPipe/32_Bytes-10                     9081            130495 ns/op          24.52 MB/s        1858 B/op        204 allocs/op
BenchmarkSyncThroughputPipe/512_Bytes-10                    8971            131535 ns/op         389.25 MB/s        1857 B/op        204 allocs/op
BenchmarkSyncThroughputPipe/1024_Bytes-10                   8842            132048 ns/op         775.47 MB/s        1858 B/op        204 allocs/op
BenchmarkSyncThroughputPipe/2048_Bytes-10                   8940            133437 ns/op        1534.81 MB/s        1859 B/op        204 allocs/op
BenchmarkSyncThroughputPipe/4096_Bytes-10                   8719            134846 ns/op        3037.53 MB/s        1863 B/op        204 allocs/op
BenchmarkSyncThroughputNetwork/32_Bytes-10                  3174            374365 ns/op           8.55 MB/s        1857 B/op        204 allocs/op
BenchmarkSyncThroughputNetwork/512_Bytes-10                 2911            385986 ns/op         132.65 MB/s        1858 B/op        204 allocs/op
BenchmarkSyncThroughputNetwork/1024_Bytes-10                2793            390751 ns/op         262.06 MB/s        1858 B/op        204 allocs/op
BenchmarkSyncThroughputNetwork/2048_Bytes-10                2764            403908 ns/op         507.05 MB/s        1867 B/op        204 allocs/op
BenchmarkSyncThroughputNetwork/4096_Bytes-10                2827            431522 ns/op         949.20 MB/s        1874 B/op        204 allocs/op
BenchmarkAsyncThroughputLarge/1MB-10                         154           7670961 ns/op        13669.42 MB/s      76954 B/op          4 allocs/op
BenchmarkAsyncThroughputLarge/2MB-10                          80          15933260 ns/op        13162.10 MB/s     249059 B/op          5 allocs/op
BenchmarkAsyncThroughputLarge/4MB-10                          24          56874384 ns/op        7374.68 MB/s     2021804 B/op          8 allocs/op
BenchmarkAsyncThroughputLarge/8MB-10                           9         119697657 ns/op        7008.16 MB/s     5165684 B/op         10 allocs/op
BenchmarkAsyncThroughputLarge/16MB-10                          5         248313308 ns/op        6756.47 MB/s    19388118 B/op         18 allocs/op
BenchmarkSyncThroughputLarge/1MB-10                           97          12269260 ns/op        8546.37 MB/s     1947469 B/op        215 allocs/op
BenchmarkSyncThroughputLarge/2MB-10                           45          23538144 ns/op        8909.59 MB/s     7063124 B/op        227 allocs/op
BenchmarkSyncThroughputLarge/4MB-10                           24          46106377 ns/op        9097.01 MB/s    21112199 B/op        241 allocs/op
BenchmarkSyncThroughputLarge/8MB-10                           10         105036533 ns/op        7986.37 MB/s    16683982 B/op        220 allocs/op
BenchmarkSyncThroughputLarge/16MB-10                           5         239043008 ns/op        7018.49 MB/s    143693908 B/op       265 allocs/op
BenchmarkTCPThroughput/32_Bytes-10                         22310             53548 ns/op          59.76 MB/s         304 B/op          4 allocs/op
BenchmarkTCPThroughput/512_Bytes-10                        14035             85581 ns/op         598.27 MB/s         304 B/op          4 allocs/op
BenchmarkTCPThroughput/1024_Bytes-10                        9780            126017 ns/op         812.59 MB/s         304 B/op          4 allocs/op
BenchmarkTCPThroughput/2048_Bytes-10                        5702            212654 ns/op         963.07 MB/s         304 B/op          4 allocs/op
BenchmarkTCPThroughput/4096_Bytes-10                        3081            394373 ns/op        1038.61 MB/s         304 B/op          4 allocs/op
BenchmarkTCPThroughput/1MB-10                                 90          12850417 ns/op        8159.86 MB/s         304 B/op          4 allocs/op
BenchmarkTCPThroughput/2MB-10                                 46          24724914 ns/op        8481.94 MB/s         304 B/op          4 allocs/op
BenchmarkTCPThroughput/4MB-10                                 21          49151189 ns/op        8533.47 MB/s         304 B/op          4 allocs/op
BenchmarkTCPThroughput/8MB-10                                 10         108572725 ns/op        7726.26 MB/s         304 B/op          4 allocs/op
BenchmarkTCPThroughput/16MB-10                                 5         232011225 ns/op        7231.21 MB/s         304 B/op          4 allocs/op
PASS
ok      github.com/loopholelabs/frisbee 86.818s
goos: darwin
goarch: arm64
pkg: github.com/loopholelabs/frisbee/pkg/metadata
BenchmarkEncodeHandler-10               200520098                5.868 ns/op
BenchmarkDecodeHandler-10               251559164                4.780 ns/op
BenchmarkEncodeDecodeHandler-10         97925209                13.83 ns/op
BenchmarkEncode-10                      226155655                5.333 ns/op
BenchmarkDecode-10                      313020477                3.815 ns/op
BenchmarkEncodeDecode-10                100000000               11.25 ns/op
PASS
ok      github.com/loopholelabs/frisbee/pkg/metadata    9.449s
PASS
ok      github.com/loopholelabs/frisbee/pkg/packet      0.095s
PASS
ok      github.com/loopholelabs/frisbee/pkg/queue       0.090s
```

## Linting

Please make sure you've run the following and fixed any issues that arise:

- [x] `trunk check` has been run

## Final Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by ShivanshVij** [comment]
Only thing left to do is change the queue implementation to use golang generics instead of unsafe.Pointer so that we can return dropped packets to the golang pool properly.

File: pkg/packet/pool.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+type Pool struct {
+	pool sync.Pool
+}
+
+func NewPool() *Pool {
+	return new(Pool)
+}
+
+func (p *Pool) Get() (s *Packet) {
```
Comment (Jspsun) [inline pkg/packet/pool.go:36]: nit: We should stick to consistently using one of either named or unamed returns throughout our codebase

File: pkg/queue/queue.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	"runtime"
+	"sync/atomic"
+	"unsafe"
+)
+
+var (
+	Closed = errors.New("queue is closed")
+)
+
+func round(value uint64) uint64 {
```
Comment (Jspsun) [inline pkg/queue/queue.go:31]: nit: this file in particular isnt very readable. We should add comments to speed up future code reviews as well as any work that touches these files

**Review by Jspsun — APPROVED** [review]

File: pkg/packet/pool.go (hunk 1 above)
Comment (ShivanshVij) [reply pkg/packet/pool.go:36]: It's mostly being done this way as a compiler hint for better performance: https://blog.min.io/golang-internals-part-2-nice-benefits-of-named-return-values-2/

**Pull request merged**